api/v1beta1/cluster_types.go

```diff
@@ -40,6 +40,238 @@ const (
+	ClusterControlPlaneInitializedV1Beta2Condition = "ControlPlaneInitialized"
+
+	// ClusterControlPlaneInitializedV1Beta2Reason surfaces when the cluster control plane is initialized.
+	ClusterControlPlaneInitializedV1Beta2Reason = "Initialized"
+
+	// ClusterControlPlaneNotInitializedV1Beta2Reason surfaces when the cluster control plane is not yet initialized.
+	ClusterControlPlaneNotInitializedV1Beta2Reason = "NotInitialized"
+
+	// ClusterControlPlaneInitializedInternalErrorV1Beta2Reason surfaces unexpected failures when computing the
+	// ControlPlaneInitialized condition.
+	ClusterControlPlaneInitializedInternalErrorV1Beta2Reason = InternalErrorV1Beta2Reason
+)
+
+// Cluster's ControlPlaneAvailable condition and corresponding reasons that will be used in v1Beta2 API version.
+const (
+	// ClusterControlPlaneAvailableV1Beta2Condition is a mirror of Cluster's control plane Available condition.
+	ClusterControlPlaneAvailableV1Beta2Condition = "ControlPlaneAvailable"
+
+	// ClusterControlPlaneInvalidConditionReportedV1Beta2Reason surfaces a control plane Available condition (read from a control plane object) which is invalid.
+	// (e.g. its status is missing).
+	ClusterControlPlaneInvalidConditionReportedV1Beta2Reason = InvalidConditionReportedV1Beta2Reason
+
+	// ClusterControlPlaneAvailableNoReasonReportedV1Beta2Reason applies to a control plane Available condition (read from a control plane object) that reports no reason.
+	ClusterControlPlaneAvailableNoReasonReportedV1Beta2Reason = NoReasonReportedV1Beta2Reason
+
+	// ClusterControlPlaneInternalErrorV1Beta2Reason surfaces unexpected failures when reading a control plane object.
+	ClusterControlPlaneInternalErrorV1Beta2Reason = InternalErrorV1Beta2Reason
+
+	// ClusterControlPlaneDoesNotExistV1Beta2Reason surfaces when a referenced control plane object does not exist.
+	// Note: this could happen when creating the Cluster. However, this state should be treated as an error if it lasts indefinitely.
+	ClusterControlPlaneDoesNotExistV1Beta2Reason = ObjectDoesNotExistV1Beta2Reason
+
+	// ClusterControlPlaneDeletedV1Beta2Reason surfaces when a referenced control plane object has been deleted.
+	// Note: controllers can't identify if the control plane object was deleted by the controller itself, e.g.
+	// during the deletion workflow, or by a users.
+	ClusterControlPlaneDeletedV1Beta2Reason = ObjectDeletedV1Beta2Reason
+)
+
+// Cluster's WorkersAvailable condition and corresponding reasons that will be used in v1Beta2 API version.
+const (
+	// ClusterWorkersAvailableV1Beta2Condition is the summary of MachineDeployment and MachinePool's Available conditions.
+	// Note: Stand-alone MachineSets and stand-alone Machines are not included in this condition.
+	ClusterWorkersAvailableV1Beta2Condition = "WorkersAvailable"
+
+	// ClusterWorkersAvailableNoWorkersV1Beta2Reason surfaces when no MachineDeployment and MachinePool exist for the Cluster.
```

Comment:
Or they have no replicas?

Reply:
ClusterWorkersAvailableNoWorkersV1Beta2Reason reason is used only of there are no objects to use from inferring workers Available.

If there is at least one MachineDeployment/MachinePool, this condition will aggregate Available condition from those objects, or, if the available conditions is not yet reported by those objects, it will consider available as unknown (and report this).

If you look at this from another angle, WorkersAvailable at cluster level should not look at replicas directly, but it should rely on whatever definition of Available is surfaced by other objects.